src/rebuild/scan.c

```diff
@@ -699,6 +717,16 @@ rebuild_scan_leader(void *data)
 	}
 	ABT_mutex_unlock(rpt->rt_lock);
 
+	/** Re-sync uncommitted DTX before rebuilding objects. */
+	rc = dtx_resync(DAOS_HDL_INVAL, rpt->rt_pool_uuid, rpt->rt_coh_uuid,
+			rpt->rt_rebuild_ver, true);
```

Comment:
Hmm, I also saw this in ds_cont_local_open()? which one really does the work?

Btw: we discussed to move this dtx rebuild things outside of rebuild, and do it separately before the rebuild. It is done in the following patch?

Reply:
The dtx_resync() is the entry for rebuild DTX status. That will be triggered when open the container normally or just before the object-rebuild.

Reply:
so you mean we need to do it in both of these two funcs? Could you please explain for container open?

Reply:
For performance consideration, we will not commit DTX one by one when they are ready for commit. Instead, we cache the committable DTXs in DRAM on leader, we call it as CoS cache. But if the leader restart, then the DTXs in the CoS cache will be lost, which on-disk status are still 'prepared'. There will be two cases for the subsequent handling:

1. The old leader is evicted and a new leader is elected, then it will trigger rebuild. Before the object rebuild, the new leader will rebuild DTX globally.

2. The old leader is NOT evicted, it starts again. Then when client tries to access some data on it, it needs to re-sync related DTXs that are committable but in 'prepared' status. That is the DTX resync when open the container.